jcl/src/java.base/share/classes/java/lang/String.java

```diff
@@ -3891,61 +3879,134 @@ static void checkBoundsBeginEnd(int begin, int end, int length) {
 		}
 		throw newStringIndexOutOfBoundsException(begin, end, length);
 	}
-	
-	private static final class StringSpliteratorOfInt implements Spliterator.OfInt {
-		private final int start; // index of first character in the array
-		private int end; // index after the final character
-		private int cursor; // next character to read
-		private final IntUnaryOperator func;
-		
-		StringSpliteratorOfInt(int start, int end, IntUnaryOperator func) {
-			this.start = start;
-			this.end = end;
-			this.func = func;
-			cursor = start;
-		}
 
-		@Override
-		public long estimateSize() {
-			return end - start;
-		}
+	@Override
+	public IntStream chars() {
+		final class CharSpliterator implements Spliterator.OfInt {
+			private final int start; // index of the first character
+			private int end; // index after the final character
+			private int cursor; // next character to read
 
-		@Override
-		public int characteristics() {
-			return IMMUTABLE | ORDERED | SIZED | SUBSIZED;
-		}
+			CharSpliterator(int start, int end) {
+				this.start = start;
+				this.end = end;
+				this.cursor = start;
+			}
+
+			@Override
+			public long estimateSize() {
+				return end - start;
+			}
 
-		@Override
-		public OfInt trySplit() {
-			int newEnd = cursor + (end - cursor + 1) / 2;
-			StringSpliteratorOfInt newSpliterator = null;
-			if (newEnd < end) {
-				newSpliterator = new StringSpliteratorOfInt(newEnd, end, func);
+			@Override
+			public int characteristics() {
+				return IMMUTABLE | ORDERED | SIZED | SUBSIZED;
+			}
+
+			@Override
+			public OfInt trySplit() {
+				int oldEnd = end;
+				int newEnd = cursor + (oldEnd - cursor + 1) / 2;
+				if (newEnd >= oldEnd) {
+					return null;
+				}
 				end = newEnd;
+				return new CharSpliterator(newEnd, oldEnd);
 			}
-			return newSpliterator;
-		}
 
-		@Override
-		public boolean tryAdvance(IntConsumer action) {
-			boolean result = false;
-			if (cursor < end) {
-				action.accept(func.applyAsInt(cursor));
-				result = true;
-				++cursor;
+			@Override
+			public boolean tryAdvance(IntConsumer action) {
+				if (cursor >= end) {
+					return false;
+				}
+				action.accept(charAtInternal(cursor));
+				cursor += 1;
+				return true;
 			}
-			return result;
 		}
+
+		return StreamSupport.intStream(new CharSpliterator(0, lengthInternal()), false);
 	}
-	
-	@Override
-	public IntStream chars() {
-		return StreamSupport.intStream(new StringSpliteratorOfInt(0, lengthInternal(), i -> charAt(i)), false);
-	}
-	
+
 	@Override
 	public IntStream codePoints() {
-		return StreamSupport.intStream(new StringSpliteratorOfInt(0, lengthInternal(), i -> codePointAt(i)), false);
+		final class CodeSpliterator implements Spliterator.OfInt {
```

Comment:
This seems effectively a duplicate of `java/lang/StringLatin1.CharsSpliterator` [1]. There is also a `StringUTF16` version as well. I think we should be able to reuse these classes rather than rolling out our own. Can we check if this is possible?

[1] https://github.com/ibmruntimes/openj9-openjdk-jdk11/blob/007fb486206c01105c4652c4f7c1c69e1ee4a624/src/java.base/share/classes/java/lang/StringLatin1.java#L737-L792

Reply:
Similarly for `chars()` as well.

Reply:
Another good suggestion: it seems to work just fine.